modules/compression/src/main/java/org/apache/ignite/spi/communication/DeflaterEngine.java

```diff
@@ -0,0 +1,189 @@
+    private static final int batchSize = 1 << 10;
+
+    /** */
+    private final Deflater deflater = new Deflater();
+
+    /** */
+    private byte[] inputDeflaterArr = new byte[initArrSize];
+
+    /** */
+    private final ExtendedByteArrayOutputStream deflateBaos = new ExtendedByteArrayOutputStream(initArrSize);
+
+    /** */
+    private final Inflater inflater = new Inflater();
+
+    /** */
+    private byte[] inputInflaterArr = new byte[initArrSize];
+
+    /** */
+    private final ExtendedByteArrayOutputStream inflateBaos = new ExtendedByteArrayOutputStream(initArrSize);
+
+    /** */
+    private int inputUnwrapPos = 0;
+
+    /** */
+    private int inputUnwrapLen = 0;
+
+    /** */
+    public DeflaterEngine(){
+        deflater.setLevel(Deflater.BEST_SPEED);
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompressionEngineResult compress(ByteBuffer src, ByteBuffer buf) {
+        assert src != null;
+        assert buf != null;
+
+        int len = src.remaining();
+
+        if (inputDeflaterArr.length < len)
+            inputDeflaterArr = new byte[U.ceilPow2(inputDeflaterArr.length)];
+
+        src.get(inputDeflaterArr, 0, len);
+
+        deflater.reset();
+        deflater.setInput(inputDeflaterArr, 0, len);
+        deflater.finish();
+
+        deflateBaos.reset();
+
+        byte[] arr = new byte[batchSize];
+
+        while (!deflater.finished()) {
+            int cnt = deflater.deflate(arr);
+
+            deflateBaos.write(arr, 0, cnt);
+        }
+
+        if (deflateBaos.size() > buf.remaining()) {
+            src.rewind();
+
+            return BUFFER_OVERFLOW;
+        }
+
+        buf.put(deflateBaos.getByteArray(), 0, deflateBaos.size());
+
+        return OK;
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompressionEngineResult decompress(ByteBuffer src, ByteBuffer buf) throws IOException {
+        assert src != null;
+        assert buf != null;
+
+        if (inflateBaos.size() > 0){
+            if (buf.remaining() < inflateBaos.size())
+                return BUFFER_OVERFLOW;
+
+            buf.put(inflateBaos.getByteArray(), 0, inflateBaos.size());
+
+            inflateBaos.reset();
+        }
+
+        int len = src.remaining();
+
+        while (inputInflaterArr.length < src.remaining())
+            inputInflaterArr = new byte[inputInflaterArr.length * 2];
+
+        if (inputUnwrapPos >= inputUnwrapLen) {
+            if (len > 0) {
+                src.get(inputInflaterArr, 0, len);
+
+                inputUnwrapLen = len;
+                inputUnwrapPos = 0;
+            }
+            else
+                return BUFFER_UNDERFLOW;
+        }
+
+        inflater.setInput(inputInflaterArr, inputUnwrapPos, inputUnwrapLen - inputUnwrapPos);
+
+        byte[] arr = new byte[batchSize];
+
+        while (!inflater.finished() && !inflater.needsInput()) {
+            try {
+                int cnt = inflater.inflate(arr);
+                inflateBaos.write(arr, 0, cnt);
```

Comment:
Add newline or split in one line like `inflateBaos.write(arr, 0, inflater.inflate(arr));`

Reply:
Done.